Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -65,18 +65,12 @@ final class ConnectionPool {
     func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
         let key = Key(request)
 
-        let provider: HTTP1ConnectionProvider = self.connectionProvidersLock.withLock {
-            if let existing = self.connectionProviders[key] {
-                existing.stateLock.withLock {
-                    existing.state.pending += 1
```

Comment:
What will be replacing this `pending` property?

Reply:
One way (but I don't know if that's still applicable with the refactor) is to expose the get connection action outside of the provider so that there doesn't need to be an additional property indicating wether or not there are still pending connections.
Like many other places I would have liked to be able to hold a lock for just a bit longer but IIRC that wasn't possible for the same potential deadlock issues you discussed with @weissi

Reply:
this property is replaced by `openedConnectionsCount`, we increase it every time we start a new connection and decrease it when connection goes away. This allows to achieve two things:
1. is easy to check for how many connections we have - either to check that we do not exceed the limit or if we closed everything when shutting down
2. this allows us to use `availableConnections` as a pool - we pop active connections for usage and offer them when we are done, I think it might be a liittle bit cleaner that iterating over connections to find "available". Additionally, we can implement different vending options - FIFO or LIFO

Reply:
@artemredkin  It's great `openedConnectionsCount` completely avoids all close callbacks, I think that alone may have been the cause of some of the bugs found in the flaky tests.

Note there might be a potential issue with `getConnection` (ConnectionPool.swift, L65) once the provider removal is implemented, because in:
```swift
let provider: HTTP1ConnectionProvider = self.lock.withLock {
    if let existing = self.providers[key] {
        return existing
    } else {
        let http1Provider = HTTP1ConnectionProvider(key: key, eventLoop: eventLoop,
configuration: self.configuration, pool: self)
        self.providers[key] = http1Provider
        return http1Provider
    }
}
return provider.getConnection(preference: preference)
```

The issue is that once we stop holding the lock and use `provider` (on the `return` line) there is nothing synchronising it anymore, so before we call the `getConnection` method on it it doesn't know there is a new tasks that asked it for a connection, and if at this point in time it were to be removed from the pool from another thread, we would be getting our connection on an already removed provider, that was what `pending` was attempting to avoid, but that wasn't elegant at all.

Reply:
Yep, you are correct, to handle that I've added `active: Bool` property